### PR TITLE
feat(sgid-login): general availability and multi-employment

### PR DESCRIPTION
**Pull request by KenLSM**
## Problem

Closes FRM-1432, closes FRM-1428
Doc on [notion](https://www.notion.so/opengov/sgID-Login-Integration-e114c71e1d164866b3b2f91effc33221?pvs=4)

## Solution

**Breaking Changes** 

- No - this PR is backwards compatible  

## Before & After Screenshots

| Section | Before | After |
|--------|--------|--------|
| Login Page | Screenshot 2023-10-05 at 6 04 30 PM | Screenshot 2023-10-16 at 5 21 27 PM |
| 🌱  Profile Selection Page | N/A | Screenshot 2023-10-05 at 6 13 19 PM  |
| 🌱 Profile Selection Page Errors | N/A | Screenshot 2023-10-16 at 5 22 29 PM | 

🌱 = new

## Tests

Note: See [ticket](https://linear.app/ogp/issue/FRM-1428/sgid-pocdex-login) for instructions to install the chrome extension to help in testing sgid login.

Regression
##### OTP login, logout
- [ ] Ensure that logging-in through OTP shows your dashboard
- [ ] Ensure that after logging-out will redirect you if you head to `/dashboard`

New Feature
##### sgID login flow happy flow
- [ ] Ensure that you are logged out
- [ ] Ensure that going to `/login/select-profile` brings you back to `/login`
- [ ] Ensure that going to `/dashboard` brings you back to `/login`
- [ ] Ensure that clicking `Log in with singpass app` will bring you to sgid login page where you are presented with a qrcode
- [ ] Ensure that scanning the qrcode with a valid account will bring you to `/login/select-profile` with 1 account
- [ ] Ensure that clicking on the account brings you to 

##### Prevent sgID login-jacking; expire select-profile session
- [ ] Click `Log in with singpass app`, and scan the QRCode
- [ ] Ensure that you are on the `/login/select-profile` page
- [ ] Ensure that you are still on the page when you refresh
- [ ] Stay on `/login/select-profile` for more than 5 minutes
- [ ] Ensure that you are redirect back to `/login` after 5 mins

##### No work employment admin
- [ ] Click `Log in with singpass app`, and scan the QRCode (find a profile that has no work employment)
- [ ] Ensure that you are on the `/login/select-profile` page
- [ ] Ensure that you see a dialog prompting `Singpass login isn't available to you yet`
- [ ] Ensure that there's no close button, and the dialog cannot be dismissed through backdrop press
- [ ] Ensure that clicking on CTA button redirects Admin back to `/login`

##### Multi-work employment admin
- [ ] Click `Log in with singpass app`, and scan the QRCode (find a profile that has multiple work employment)
- [ ] Ensure that you are on the `/login/select-profile` page
- [ ] Ensure that you see multiple entries of work employment
- [ ] Ensure that the selected profile is correctly logged in
- [ ] Return to `/login/select-profile`
- [ ] Ensure that clicking on a different profile through allows you to switch profile (so long it is within 5 mins)
- [ ] Ensure that the selected profile is correctly logged in
Note: profile switching is an out-of-scope feature, that is can be extended in the future

##### Non-whitelisted admin login
- [ ] Click `Log in with singpass app`, and scan the QRCode (find a profile that is not from a whitelisted domain)
- [ ] Ensure that you are on the `/login/select-profile` page
- [ ] Click on the profile
- [ ] Ensure that you see a dialog prompting `You don't have access to this service`
- [ ] Ensure that clicking on close button, and backdrop press will dismiss the dialog
- [ ] Ensure that clicking on CTA button redirects Admin back to `/login`

## Deploy Notes

- [x] `SGID_ADMIN_LOGIN_REDIRECT_URI` to be updated to `https://form.gov.sg/api/v3/auth/sgid/login` with their respectively subdomains for `staging`,`staging-alt`,`staging-alt2`, `uat`, and `prod`
- [x] Register POCDEX scope (`pocdex.public_officer_details`) on developer.id.gov.sg

### CODE REVIEW

File: frontend/src/features/login/components/SgidLoginButton.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        <Flex align="center" flexDirection="row">
+          <Text color="primary.500">Log in with </Text>
+          <SingpassFullLogoSvgr height="1.25rem" />
+          <Text color="primary.500"> app</Text>
+        </Flex>
+      </Button>
+      <Text>
+        For{' '}
+        <Link isExternal href={SGID_VALID_ORG_PAGE}>
+          select organization
```
Comment (wanlingt) [inline frontend/src/features/login/components/SgidLoginButton.tsx]: suggestion: `select agencies` instead to keep the copy consistent? the docs use 'agencies'

File: frontend/src/features/login/SelectProfilePage.tsx [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  body: string | (() => React.ReactElement)
+  cta: string
+  onCtaClick: (disclosureProps: ErrorDisclosureProps) => void
+}
+
+const MODAL_ERRORS: Record<string, ModalErrorMessages> = {
+  NO_WORKEMAIL: {
+    hideCloseButton: true,
+    header: "Singpass login isn't available to you yet",
+    body: 'It is progressively being made available to organisations. In the meantime, please log in using your email address.',
```
Comment (wanlingt) [inline frontend/src/features/login/SelectProfilePage.tsx]: same comment as above to use 'agencies' instead

File: frontend/src/features/login/LoginPage.tsx [hunk 3]
```diff
@@ -22,6 +27,27 @@ export const LoginPage = (): JSX.Element => {
...
   const [otpPrefix, setOtpPrefix] = useState<string>('')
 
+  const [params] = useSearchParams()
+  const toast = useToast({ isClosable: true, status: 'danger' })
+
+  const statusCode = params.get('status')
+  const toastMessage = useMemo(() => {
+    switch (statusCode) {
+      case null:
+      case '200':
```
Comment (wanlingt) [inline frontend/src/features/login/LoginPage.tsx]: can consider using the enum constant `StatusCodes.OK`

`import { StatusCodes } from 'http-status-codes'
`

File: frontend/src/features/login/SelectProfilePage.tsx [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    ApiService.post<void>(SGID_PROFILES_ENDPOINT, {
+      workEmail: profile.work_email,
+    })
+      .then(() => {
+        window.location.assign(DASHBOARD_ROUTE)
+        setIsAuthenticated(true)
+      })
+      .catch((err) => {
+        console.log({ err })
+        if (err.code === 401) {
```
Comment (wanlingt) [inline frontend/src/features/login/SelectProfilePage.tsx]: same comment as above, can consider using status codes constants enum

File: frontend/src/features/login/queries.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+import { ApiService } from '~services/ApiService'
+
+export const SGID_PROFILES_ENDPOINT = '/auth/sgid/profiles'
+
+export const useSgidProfiles = (): UseQueryResult<
+  SgidProfilesDto,
+  ApiError
+> => {
+  return useQuery<SgidProfilesDto, ApiError>(['sgidProfiles'], async () => {
```
Comment (wanlingt) [inline frontend/src/features/login/queries.ts]:
```suggestion

export const sgidProfilesKeys = {
  base: ['sgidProfiles'] as const,
}

...
  return useQuery<SgidProfilesDto, ApiError>(sgidProfilesKeys.base, async () => {
```

for consistency with other queries

File: src/types/vendor/express.d.ts [hunk 6]
```diff
@@ -37,6 +38,7 @@ declare module 'express-session' {
     user?: {
       _id: IUserSchema['_id']
     }
+    sgid?: SgidUser
```
Comment (wanlingt) [inline src/types/vendor/express.d.ts:41]: should we set this in `setProfile` in `auth-sgid.controller`? don't think this is being used currently (lmk if I missed it out somewhere!)

**Review by wanlingt — COMMENTED** [review]
minor comments!

File: src/types/vendor/express.d.ts (hunk 6 above)
Comment (KenLSM) [reply src/types/vendor/express.d.ts:41]: This is used in `setProfile` to verify if the selected profile is valid on [Line 220](https://github.com/opengovsg/FormSG/pull/6763/files#diff-1f3c9626dd630dc0f340b56cb7ebedabd983255deb8817f48369b3d3c321fe7cR220).

The actual "user" identity is stored under `session.user._id`, while the sgid profile information is stored under `session.sgid.profiles`.

Lmk if it's confusing.

Comment (KenLSM) [reply src/types/vendor/express.d.ts:41]: To be verbose.

1. `sgid.profiles` are valid profiles (can be multiple) that the user can assume. But the user can only assume 1 profile at any single time (identified through `user._id`.

2. I'm storing the profiles under `session` as it's a two-step process where we have a (1) callback from sgid supplying us the valid sgid profiles, and (2) where the user selects from the list of valid profiles.

File: frontend/src/features/login/SelectProfilePage.tsx (hunk 2 above)
Comment (KenLSM) [reply frontend/src/features/login/SelectProfilePage.tsx]: 293cb9839

File: frontend/src/features/login/components/SgidLoginButton.tsx (hunk 1 above)
Comment (KenLSM) [reply frontend/src/features/login/components/SgidLoginButton.tsx]: 293cb9839

File: frontend/src/features/login/SelectProfilePage.tsx (hunk 4 above)
Comment (KenLSM) [reply frontend/src/features/login/SelectProfilePage.tsx]: 67e5d1811

File: frontend/src/features/login/LoginPage.tsx (hunk 3 above)
Comment (KenLSM) [reply frontend/src/features/login/LoginPage.tsx]: Great suggestion! Looks neater now 67e5d1811

File: frontend/src/features/login/queries.ts (hunk 5 above)
Comment (KenLSM) [reply frontend/src/features/login/queries.ts]: 42010a2e0

File: src/types/vendor/express.d.ts (hunk 6 above)
Comment (wanlingt) [reply src/types/vendor/express.d.ts:41]: Ok yes, got it! I was wondering where the profiles were being set, realise it's in `handleLoginCallback`

**Review by wanlingt — APPROVED** [review]
lgtm! btw are we rolling this out at one go (instead of incrementally through growthbook)? I assume it's fine, since they can log in through the usual method if this one fails

also in the tests, can you include the NRICs of mock profiles that fit those usecases so it's easier for the tester? ty!

**Comment by KenLSM** [comment]
> lgtm! btw are we rolling this out at one go (instead of incrementally through growthbook)? I assume it's fine, since they can log in through the usual method if this one fails

There were two considerations on my side.
1. This should be relatively low-risk, and like you've identified, the users can login through the usual method.
2. I couldn't think of a way to divide the rollout into different segments deterministically (no formid, agencyid, userid, etc to use), lmk if you have any ideas

 
> also in the tests, can you include the NRICs of mock profiles that fit those usecases so it's easier for the tester? ty!

I've added them in the ticket! (I think the info is a little sensitive)

**Comment by wanlingt** [comment]
> 2. I couldn't think of a way to divide the rollout into different segments deterministically (no formid, agencyid, userid, etc to use), lmk if you have any ideas

Could use the percentage rollout on growthbook, though this is random so the user might not be shown the same login method the next time they visit the page, which probably isn't ideal. Anyway I'm on board with just rolling this out directly

**Pull request merged**